### PR TITLE
Support free response/multiple choice summary page for lab2 levels

### PR DESCRIPTION
**Pull request by molly-moen**
Teacher tools: tagging you on this PR since you own this feature. The PR should not change any existing behavior, just updates for lab2!

We have a per-level summary page for levels that are either free response/multiple choice or have a contained free response/multiple choice level. This shows a per-section summary of student responses, either as a graph for multiple choice or a list of responses. There is a button for teachers in the instructions panel that takes them to that page, along with a summary of how many students have responded.

We want to support this flow for lab2 as well, which uses new predict level configuration rather than a contained level (see linked spec for details). To do this, I updated the summary page to either use the old method of loading the settings, or to get the predict settings from the `predict_settings` object if it exists. The responses are stored in the same way for both level types, so no changes were needed there. 

I also updated how we store multiple choice responses for predict levels to match the old way, which is to store a list of the indices of the response(s) rather than the response strings.

The summary page displays the question and correct answer all via haml. For predict levels I added two divs that we inject the question and correct answer (defined as react components) into for predict levels. I was able to reuse the same component that we use to display the solution to teachers (I updated that component to have a non-redux variant), but made a new component to show the question as we don't want or need the question to have editable inputs.

Finally, I made a new button that links to the summary page from the lab2 instructions panel. Only teachers see this on predict levels. I haven't added the summary of student responses yet, I will cover that as a follow up.

### Old levels
Button in instructions
Screenshot 2024-07-09 at 1 22 00 PM

Free Response answers
Screenshot 2024-07-09 at 1 23 25 PM

Multiple Choice answers
Screenshot 2024-07-09 at 1 23 32 PM

### Lab2 version
Button in instructions
Screenshot 2024-07-09 at 1 30 47 PM

Free response
Screenshot 2024-07-09 at 1 24 17 PM

Multiple choice
Screenshot 2024-07-09 at 1 24 24 PM

## Links

- [spec](https://docs.google.com/document/d/1lrpc06mMV8CiYTFfx5ahKZ887qBEJ5AybHPn1Pd8JM0/edit#heading=h.n24aegko5o63)
- jira ticket: [CT-667](https://codedotorg.atlassian.net/browse/CT-667)

## Testing story
Tested locally that old and new predict/contained levels show the summary page correctly, and other lab2 non-predict levels are unaffected. I also updated the unit test for the Multi Response summary component.

## Follow-up work
Two follow up items:
- Show "x/y students answered" in the instructions panel. Adding this probably requires adding a new api since this information was previously added in haml. [Task](https://codedotorg.atlassian.net/browse/CT-678) to add this.
- You can navigate to and from the summary page in lab2 by clicking the button on the header, then clicking "back to level" on the summary page. However, if you click a bubble in the header on the summary page it doesn't seem like anything happens, because we are trying to change levels without a page reload but aren't actually in a lab2 level. [Task](https://codedotorg.atlassian.net/browse/CT-679) to fix this.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/levelSummary/SummaryPredictQuestion.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    <div>
+      <EnhancedSafeMarkdown markdown={question} />
+      {predictSettings.questionType === PredictQuestionType.FreeResponse ? (
+        <textarea
+          placeholder={predictSettings.placeholderText}
+          style={{height: predictSettings.freeResponseHeight || 20}}
+          readOnly={true}
+        />
+      ) : (
+        predictSettings.multipleChoiceOptions?.map((option, index) => {
```
Comment (sanchitmalhotra126) [inline apps/src/templates/levelSummary/SummaryPredictQuestion.tsx:30]: Could this be shared with `PredictQuestion` at all? I noticed there's some similar stuff with how the multiple choice answers get displayed, but also not sure if there's much beyond that.

File: apps/src/templates/levelSummary/SummaryPredictQuestion.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+interface SummaryPredictQuestionProps {
+  question: string;
+  predictSettings: LevelPredictSettings;
+}
+
+const SummaryPredictQuestion: React.FunctionComponent<
+  SummaryPredictQuestionProps
+> = ({question, predictSettings}) => {
+  console.log(predictSettings);
```
Comment (sanchitmalhotra126) [inline apps/src/templates/levelSummary/SummaryPredictQuestion.tsx]: nit: remove?

File: apps/src/templates/levelSummary/SummaryPredictQuestion.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  return (
+    <div>
+      <EnhancedSafeMarkdown markdown={question} />
+      {predictSettings.questionType === PredictQuestionType.FreeResponse ? (
+        <textarea
+          placeholder={predictSettings.placeholderText}
+          style={{height: predictSettings.freeResponseHeight || 20}}
+          readOnly={true}
```
Comment (sanchitmalhotra126) [inline apps/src/templates/levelSummary/SummaryPredictQuestion.tsx:27]: For my understanding - why do we need this readonly textarea at all? I guess just to show the placeholder text?

File: apps/src/templates/levelSummary/MultiResponses.jsx [hunk 4]
```diff
@@ -48,25 +48,53 @@ const multiChartData = (data, highlightCorrect = false) => [
 ];
 
 const MultiResponses = ({scriptData, showCorrectAnswer = false}) => {
-  const answers = scriptData.level.properties.answers;
+  // Get the number of answers and a list of correct answers.
+  // The correct answers will be the letters for the correct answers, ex. ['A', 'C']
+  const [numAnswers, correctAnswers] = useMemo(() => {
```
Comment (sanchitmalhotra126) [inline apps/src/templates/levelSummary/MultiResponses.jsx:53]: Minor: could this be consolidated to something like
```
  let answers = [];
  let correctIndices = [];

  if (scriptData.level.properties.answers) {
    answers = scriptData.level.properties.answers;
    correctIndices = answers
      .filter(answer => answer.correct)
      .map(answer => answers.indexOf(answer));

  } else if (scriptData.level_properties.predict_settings) {
    const predictSettings = scriptData.level.properties.predict_settings;
    answers = predictSettings.multipelChoiceOptions;
    correctIndices = predictSettings.solution
      .split(',')
      .map(answer => answers.indexOf(answer));
  }

  const correctAnswers = correctIndices.map(index =>
    String.fromCharCode(65 + index)
  );

  return [answers.length, correctAnswers];
```

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM overall!

Out of scope for this PR but asking the perennial Lab2 question 😄 would it be beneficial for us to have this in a new component that isn't routed through the existing `summary` HAML/JS entrypoint? I could imagine Lab2 displaying a summary view (via the LabViewsRenderer?) that displays the predict question/solution for the current level, and makes an AJAX call to fetch the summary responses, without a full page reload. We could still use the existing `SummaryResponses` component even (maybe with a bit of refactoring so we don't need all the "is predict level" checks inside the component)

**Comment by molly-moen** [comment]
> LGTM overall!
> 
> Out of scope for this PR but asking the perennial Lab2 question 😄 would it be beneficial for us to have this in a new component that isn't routed through the existing `summary` HAML/JS entrypoint? I could imagine Lab2 displaying a summary view (via the LabViewsRenderer?) that displays the predict question/solution for the current level, and makes an AJAX call to fetch the summary responses, without a full page reload. We could still use the existing `SummaryResponses` component even (maybe with a bit of refactoring so we don't need all the "is predict level" checks inside the component)

Yeah it would be better to have this all in a component rather than going back to the server, but it was just too much work to get that working right now. I filed a task to investigate this! [CT-680](https://codedotorg.atlassian.net/browse/CT-680)

File: apps/src/templates/levelSummary/SummaryPredictQuestion.tsx (hunk 2 above)
Comment (molly-moen) [reply apps/src/templates/levelSummary/SummaryPredictQuestion.tsx]: good catch 🤦‍♀️

File: apps/src/templates/levelSummary/SummaryPredictQuestion.tsx (hunk 3 above)
Comment (molly-moen) [reply apps/src/templates/levelSummary/SummaryPredictQuestion.tsx:27]: Yeah I was trying to match the existing behavior, which shows the readonly textarea. I don't have a strong opinion on whether we should show it or not.

File: apps/src/templates/levelSummary/SummaryPredictQuestion.tsx (hunk 1 above)
Comment (molly-moen) [reply apps/src/templates/levelSummary/SummaryPredictQuestion.tsx:30]: Yeah it was different enough that I didn't share code (`PredictQuestion` displays radio buttons/checkboxes for example). It didn't seem worth extracting out a helper for the one line conversion from index to letter, but I can do that if you think it's valuable!

File: apps/src/templates/levelSummary/MultiResponses.jsx (hunk 4 above)
Comment (molly-moen) [reply apps/src/templates/levelSummary/MultiResponses.jsx:53]: good idea!

File: apps/src/templates/levelSummary/SummaryPredictQuestion.tsx (hunk 1 above)
Comment (sanchitmalhotra126) [reply apps/src/templates/levelSummary/SummaryPredictQuestion.tsx:30]: Not necessary if there's not much to be shared!

**Pull request merged**